specs/vm/instruction_set.md

```diff
@@ -1386,6 +1387,53 @@ Cease VM execution and revert script effects. After a revert:
 1. All [OutputContract](../protocol/tx_format.md#outputcontract) outputs will have the same `balanceRoot` and `stateRoot` as on initialization.
 1. All [OutputVariable](../protocol/tx_format.md#outputvariable) outputs will have `to`, `amount`, and `asset_id` of zero.
 
+### SMO: Send message to output
+
+|             |                                                                                     |
+|-------------|-------------------------------------------------------------------------------------|
+| Description | Send a message to recipient address `MEM[$rA, 32]` with call abi `MEM[$rA + 32, $rB]` and `$rD` coins, with output `$rC`. |
+| Operation   | ```outputmessage(MEM[$fp, 32], MEM[$rA, 32], MEM[$rA + 32, $rB], $rD, $rC);```      |
+| Syntax      | `smo $rA, $rB, $rC, $rD`                                                            |
+| Encoding    | `0x00 rA rB rC rD`                                                                  |
+| Notes       |                                                                                     |
+
+Given helper `balanceOfStart(asset_id: byte[32]) -> uint32` which returns the memory address of the remaining free balance of `asset_id`, or panics if `asset_id` has no free balance remaining.
+
+Panic if:
+
+- `$rA + 32` overflows
+- `$rA + $rB + 32` overflows
+- `$rA + 32 > VM_MAX_RAM`
+- `$rA + $rB + 32 > VM_MAX_RAM`
+- `$rC > tx.outputsCount`
+- In an external context, if `$rD > MEM[balanceOfStart(0), 8]`
+- In an internal context, if `$rD` is greater than the balance of asset ID 0 of output with contract ID `MEM[$fp, 32]`
+- `tx.outputs[$rC].type != OutputType.Message`
+- `tx.outputs[$rC].recipient != 0 || tx.outputs[$rC].sender != 0`
+
+Append a receipt to the list of receipts, modifying `tx.receiptsRoot`:
+
+| name         | type          | description                                                               |
+|--------------|---------------|---------------------------------------------------------------------------|
+| `type`       | `ReceiptType` | `ReceiptType.MessageOut`                                                  |
+| `messageID`  | `byte[32]`    | The messageID as described [here](../protocol/identifiers.md#message-id). |
+| `sender`     | `byte[32]`    | The address of the message sender: `MEM[$fp, 32]`.                        |
+| `recipient`  | `byte[32]`    | The address of the message recipient: `MEM[$rA, 32]`.                     |
+| `dataLength` | `uint16`      | Length of message data, in bytes: `$rB - ($rA + 32)`.                     |
+| `amount`     | `uint64`      | Amount of base asset coins sent with message: `$rD`.                      |
+| `nonce`      | `byte[32]`    | The message nonce.                                                        |
+| `data`       | `byte[]`      | The message data or [abi encoded](https://docs.soliditylang.org/en/v0.8.13/abi-spec.html) call to execute: `MEM[$rA + 32, $rB]`.        |
```

Comment:
Rather than `data`, this should be `digest`, similar to the `LOGD` and `RETD` receipts. Essentially, this guarantees that computing the receipt hash is constant time. The actual data is collected by the implementation and available in the returned receipt implementations-side. For example: https://github.com/FuelLabs/fuel-specs/blob/fcfe18a585a465c750fd939b0fd85a30255d6c00/specs/protocol/abi.md#logdata-receipt

Reply:
The `LOGD` and `RETD` include both a `digest` and the full `data`. I'll be sure to add a `digest` to the receipt, but I'm confused as to what you mean by "receipt hash"

Reply:
To clarify: the _canonical_ `LOGD` receipt (the thing that's in the instruction set file, here) only includes a digest (similar for `RETD`): https://github.com/FuelLabs/fuel-specs/blob/fcfe18a585a465c750fd939b0fd85a30255d6c00/specs/vm/instruction_set.md#logd-log-data-event. The _JSON_ `LOGD` receipt includes both the digest and data (i.e. preimage to the digest): https://github.com/FuelLabs/fuel-specs/blob/fcfe18a585a465c750fd939b0fd85a30255d6c00/specs/protocol/abi.md#logdata-receipt

By "receipt hash" I actually meant "hashing a leaf in the receipts tree." If leaves only contain digests (they do), then they're constant-sized and that makes it easier to do some fraud proof cases and also serve light client proofs. Sometimes you don't care about the data but only that a receipt happened. If the data isn't part of the canonical receipt then you can still get a proof to the canonical receipt without having to download all the data.